### PR TITLE
feat: Websocket connection uptake

### PR DESCRIPTION
**Pull request by purnimavenkatasubbu**
**- What I did**
Implemented WebSocket support for inbound connections in the secondary server. Added InboundWebSocketConnection to handle WebSocket-based client connections.

**- How I did it**

- Created the InboundWebSocketConnection class, implementing the InboundConnection interface to handle WebSocket operations like writing, closing, and managing metadata.
- Updated AtConnectionFactory to support creating WebSocket connections via createWebSocketConnection.
- Modified InboundConnectionManager to manage WebSocket connections, ensuring proper pooling and handling of capacity limits.
- Added webSocketListener in AtSecondaryImpl to accept and handle WebSocket requests.
- Modified end-to-end test utilities and made the end2end12 job use websocket connections

**- How to verify it**
All tests pass

### CODE REVIEW

File: packages/at_secondary_server/CHANGELOG.md [hunk 1]
```diff
@@ -1,3 +1,5 @@
+# 3.1.2
```
Comment (gkc) [inline packages/at_secondary_server/CHANGELOG.md]: 3.2.0

File: packages/at_secondary_server/pubspec.yaml [hunk 2]
```diff
@@ -1,6 +1,6 @@
 name: at_secondary
 description: Implementation of secondary server.
-version: 3.1.1
+version: 3.1.2
```
Comment (gkc) [inline packages/at_secondary_server/pubspec.yaml]: 3.2.0 (new feature)

File: .github/workflows/at_server.yaml [hunk 3]
```diff
@@ -569,12 +569,14 @@ jobs:
               sed -i "s/ATSIGN_1_NAME/@$AT_SIGN_1/g" tests/at_end2end_test/config/config-e2e_test_runtime.yaml
               sed -i "s/ATSIGN_1_PORT/$AT_SIGN_1_PORT/g" tests/at_end2end_test/config/config-e2e_test_runtime.yaml
               sed -i "s/ATSIGN_1_HOST/$AT_SIGN_1_HOST/g" tests/at_end2end_test/config/config-e2e_test_runtime.yaml
+#             TODO: add something like `sed -i "s/ATSIGN_1_CONNECTION_TYPE/$AT_SIGN_1_CONNECTION_TYPE/g" tests/at_end2end_test/config/config-e2e_test_runtime.yaml`
```
Comment (gkc) [inline .github/workflows/at_server.yaml:572]: @purnimavenkatasubbu @VJag

File: .github/workflows/at_server.yaml [hunk 4]
```diff
@@ -569,12 +569,14 @@ jobs:
...
               sed -i "s/ATSIGN_2_NAME/@$AT_SIGN_2/g" tests/at_end2end_test/config/config-e2e_test_runtime.yaml
               sed -i "s/ATSIGN_2_PORT/$AT_SIGN_2_PORT/g" tests/at_end2end_test/config/config-e2e_test_runtime.yaml
-              sed -i "s/ATSIGN_2_HOST/$AT_SIGN_2_HOST/g" tests/at_end2end_test/config/config-e2e_test_runtime.yaml  
+              sed -i "s/ATSIGN_2_HOST/$AT_SIGN_2_HOST/g" tests/at_end2end_test/config/config-e2e_test_runtime.yaml
+#             TODO: add something like `sed -i "s/ATSIGN_2_CONNECTION_TYPE/$AT_SIGN_2_CONNECTION_TYPE/g" tests/at_end2end_test/config/config-e2e_test_runtime.yaml`
```
Comment (gkc) [inline .github/workflows/at_server.yaml:576]: @purnimavenkatasubbu @VJag

File: tests/at_end2end_test/config/config-e2e_test_runtime.yaml [hunk 5]
```diff
@@ -12,10 +12,12 @@ first_atsign_server:
   first_atsign_port: ATSIGN_1_PORT
   # # The url to connect to first atServer
   first_atsign_url: ATSIGN_1_HOST
+  # TODO add something like `first_atsign_connection_type: ATSIGN_1_CONNECTION_TYPE`
```
Comment (gkc) [inline tests/at_end2end_test/config/config-e2e_test_runtime.yaml]: @purnimavenkatasubbu @VJag

File: tests/at_end2end_test/config/config-e2e_test_runtime.yaml [hunk 6]
```diff
@@ -12,10 +12,12 @@ first_atsign_server:
...
 
 second_atsign_server:
   second_atsign_name: 'ATSIGN_2_NAME'
   # # The port to connect to second atServer
   second_atsign_port: ATSIGN_2_PORT
   # # The url to connect to second atServer
   second_atsign_url: ATSIGN_2_HOST
+  # TODO add something like `second_atsign_connection_type: ATSIGN_2_CONNECTION_TYPE`
```
Comment (gkc) [inline tests/at_end2end_test/config/config-e2e_test_runtime.yaml]: @purnimavenkatasubbu @VJag

File: tests/at_end2end_test/test/e2e_test_utils.dart [hunk 7]
```diff
@@ -28,7 +28,11 @@ Future<SimpleOutboundSocketHandler> getSocketHandler(atSign) async {
   if (asc == null) {
     throw _NoSuchAtSignException('$atSign not configured');
   }
+  // TODO switch on _AtSignConfig.connectionType and create a
+  // TODO SimpleOutboundSocketConnection or SimpleOutboundWebsocketConnection
+  // TODO as required
```
Comment (gkc) [inline tests/at_end2end_test/test/e2e_test_utils.dart:33]: @purnimavenkatasubbu @VJag

File: tests/at_end2end_test/test/e2e_test_utils.dart [hunk 8]
```diff
@@ -38,6 +42,9 @@ Future<SimpleOutboundSocketHandler> getSocketHandler(atSign) async {
 
 /// A simple wrapper around a socket for @ protocol communication.
 class SimpleOutboundSocketHandler {
+  // TODO Turn this into an abstract base class, e.g. SimpleOutboundConnection
+  // TODO with two concrete subclasses, e.g. SimpleOutboundSocketConnection
+  // TODO and SimpleOutboundWebsocketConnection
```
Comment (gkc) [inline tests/at_end2end_test/test/e2e_test_utils.dart]: @purnimavenkatasubbu @VJag

File: tests/at_end2end_test/test/e2e_test_utils.dart [hunk 9]
```diff
@@ -195,14 +202,22 @@ class SimpleOutboundSocketHandler {
     }
   }
 }
+
+enum _ConnectionTypeEnum {
+  socket,
+  websocket,
+}
```
Comment (gkc) [inline tests/at_end2end_test/test/e2e_test_utils.dart:317]: @purnimavenkatasubbu @VJag added an enum

File: tests/at_end2end_test/test/e2e_test_utils.dart [hunk 10]
```diff
@@ -195,14 +202,22 @@ class SimpleOutboundSocketHandler {
...
   String atSign;
   String host;
   int port;
+  _ConnectionTypeEnum connectionType;
+  // TODO Add connectionType
 
   /// Creates and adds to [atSignConfigMap] or throws [_AtSignAlreadyAddedException] if we've already got it.
-  _AtSignConfig(this.atSign, this.host, this.port) {
+  _AtSignConfig(this.atSign, this.host, this.port, this.connectionType) {
+    this.connectionType ??= _ConnectionTypeEnum.socket;
```
Comment (gkc) [inline tests/at_end2end_test/test/e2e_test_utils.dart]: if null, default to `socket`

File: tests/at_end2end_test/test/e2e_test_utils.dart [hunk 11]
```diff
@@ -232,12 +247,16 @@ void _loadTheYaml() {
   _AtSignConfig(
       ConfigUtil.getYaml()!['first_atsign_server']['first_atsign_name'],
       ConfigUtil.getYaml()!['first_atsign_server']['first_atsign_url'],
-      ConfigUtil.getYaml()!['first_atsign_server']['first_atsign_port']);
+      ConfigUtil.getYaml()!['first_atsign_server']['first_atsign_port'],
+      ConfigUtil.getYaml()!['first_atsign_server']['first_atsign_connection_type'],
```
Comment (gkc) [inline tests/at_end2end_test/test/e2e_test_utils.dart]: pass the connectionType to the modified constructor

File: tests/at_end2end_test/test/e2e_test_utils.dart [hunk 12]
```diff
@@ -28,10 +28,17 @@ Future<SimpleOutboundSocketHandler> getSocketHandler(atSign) async {
...
+  // ignore: prefer_typing_uninitialized_variables
+  var handler;
+  switch(asc.connectionType!) {
+    case _ConnectionTypeEnum.socket:
+      handler = SimpleOutboundSocketHandler._(asc.host, asc.port, atSign);
+      break;
+    case _ConnectionTypeEnum.websocket:
+      // TODO e.g. handler = SimpleOutboundWebsocketConnection._(asc.host, asc.port, atSign);
+      throw UnimplementedError('e2e_test_utils cannot yet create a websocket connection');
+  }
```
Comment (gkc) [inline tests/at_end2end_test/test/e2e_test_utils.dart:49]: @purnimavenkatasubbu @VJag

File: .github/workflows/at_server.yaml [hunk 13]
```diff
@@ -569,11 +569,11 @@ jobs:
               sed -i "s/ATSIGN_1_NAME/@$AT_SIGN_1/g" tests/at_end2end_test/config/config-e2e_test_runtime.yaml
               sed -i "s/ATSIGN_1_PORT/$AT_SIGN_1_PORT/g" tests/at_end2end_test/config/config-e2e_test_runtime.yaml
               sed -i "s/ATSIGN_1_HOST/$AT_SIGN_1_HOST/g" tests/at_end2end_test/config/config-e2e_test_runtime.yaml
-#             TODO: add something like `sed -i "s/ATSIGN_1_CONNECTION_TYPE/$AT_SIGN_1_CONNECTION_TYPE/g" tests/at_end2end_test/config/config-e2e_test_runtime.yaml`
+              # TODO: add something like sed -i "s/ATSIGN_1_CONNECTION_TYPE/$AT_SIGN_1_CONNECTION_TYPE/g" tests/at_end2end_test/config/config-e2e_test_runtime.yaml
```
Comment (gkc) [inline .github/workflows/at_server.yaml:572]: @purnimavenkatasubbu @VJag

**Comment by gkc** [comment]
@purnimavenkatasubbu @VJag I've pushed some changes with some TODOs for you to get all of these changes covered by end to end tests. Once that's all done you could consider doing the same type of thing for the functional tests and running functional tests four times (dart stable socket, dart stable websocket, dart latest socket, dart latest websocket) instead of the current two (dart stable, dart latest)

File: packages/at_secondary_server/CHANGELOG.md (hunk 1 above)
Comment (purnimavenkatasubbu) [reply packages/at_secondary_server/CHANGELOG.md]: Updated

File: packages/at_secondary_server/pubspec.yaml (hunk 2 above)
Comment (purnimavenkatasubbu) [reply packages/at_secondary_server/pubspec.yaml]: Updated

File: tests/at_end2end_test/config/config-e2e_test_runtime.yaml (hunk 5 above)
Comment (purnimavenkatasubbu) [reply tests/at_end2end_test/config/config-e2e_test_runtime.yaml]: Updated

File: tests/at_end2end_test/config/config-e2e_test_runtime.yaml (hunk 6 above)
Comment (purnimavenkatasubbu) [reply tests/at_end2end_test/config/config-e2e_test_runtime.yaml]: Updated

File: tests/at_end2end_test/test/e2e_test_utils.dart (hunk 8 above)
Comment (purnimavenkatasubbu) [reply tests/at_end2end_test/test/e2e_test_utils.dart]: Updated

File: tests/at_end2end_test/test/e2e_test_utils.dart (hunk 12 above)
Comment (purnimavenkatasubbu) [reply tests/at_end2end_test/test/e2e_test_utils.dart:49]: Updated

File: tests/at_end2end_test/config/config-e2e_test_runtime.yaml [hunk 14]
```diff
@@ -12,10 +12,14 @@ first_atsign_server:
   first_atsign_port: ATSIGN_1_PORT
   # # The url to connect to first atServer
   first_atsign_url: ATSIGN_1_HOST
+  # # The connection type of first atServer
+  first_atsign_connection_type: ATSIGN_1_CONNECTION_TYPE
```
Comment (gkc) [inline tests/at_end2end_test/config/config-e2e_test_runtime.yaml]: Rollback the changes to this file

**Review by gkc — CHANGES_REQUESTED** [review]

File: tests/at_end2end_test/config/config-e2e_test_runtime.yaml (hunk 14 above)
Comment (purnimavenkatasubbu) [reply tests/at_end2end_test/config/config-e2e_test_runtime.yaml]: Updated

**Review by gkc — APPROVED** [review]

**Pull request merged**